### PR TITLE
Vendor in Libnetwork changes

### LINKED ISSUE #41822: Upgrade to 20.10 fails to release ports published to IPv6 literal addresses
**Description**

Upon upgrading to Docker `20.10.1` any container which has a port published to a hosts IPv6 literal address cannot clean up after itself. Initial launch of the container is successful but any subsequent restarts will result in an error.
Restarting the Docker daemon will cause the `docker-proxy` instance holding the port open to be released however the issue will reoccur upon attempts to restart the container again. 

**Steps to reproduce the issue:**
1. `docker run -d --name nginx-test -p [fc00:1337:1337:1337:1337:1337:1337:1337]:80:80 nginx`
2. `docker restart nginx-test`

**Describe the results you received:**

Once issuing the restart the following error is received:
`Error response from daemon: Cannot restart container nginx-test: driver failed programming external connectivity on endpoint nginx-test (9a6f7a503369a88aa1d200050e4a82ba09e755e48588f4000086c5abae6c3710): Bind for fc00:1337:1337:1337:1337:1337:1337:1337:80 failed: port is already allocated`

Upon checking the status of the docker daemon via systemctl I can see there's an instance of `docker-proxy` holding on to the port:

```
systemctl status docker
● docker.service - Docker Application Container Engine
     Loaded: loaded (/usr/lib/systemd/system/docker.service; enabled; vendor preset: disabled)
     Active: active (running) since Thu 2020-12-17 13:04:12 AEDT; 1 day 23h ago
TriggeredBy: ● docker.socket
       Docs: https://docs.docker.com
   Main PID: 384483 (dockerd)
      Tasks: 41 (limit: 38458)
     Memory: 265.4M
     CGroup: /system.slice/docker.service
             ├─384483 /usr/bin/dockerd -H fd://
             ├─384495 containerd --config /var/run/docker/containerd/containerd.toml --log-level info
             └─634622 /usr/bin/docker-proxy -proto tcp -host-ip fc00:1337:1337:1337:1337:1337:1337:1337 -host-port 80 -container-ip 172.17.0.2 -container-port 80
```

The logs for the Docker daemon also provide the following output:

```
Dec 19 12:04:11 mana dockerd[384483]: time="2020-12-19T12:04:11.366530647+11:00" level=info msg="ignoring event" container=ed24834d6f45bbaf301c8243f51a942bb82d81675c70481f68d64cf14f8f005e module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Dec 19 12:04:11 mana dockerd[384495]: time="2020-12-19T12:04:11.366851423+11:00" level=info msg="shim disconnected" id=ed24834d6f45bbaf301c8243f51a942bb82d81675c70481f68d64cf14f8f005e
Dec 19 12:04:11 mana dockerd[384483]: time="2020-12-19T12:04:11.368999767+11:00" level=warning msg="\ncould not release {tcp 172.17.0.2 80 fc00:1337:1337:1337:1337:1337:1337:1337 80 80} because of port is not mapped"
Dec 19 12:04:11 mana dockerd[384483]: time="2020-12-19T12:04:11.582843296+11:00" level=warning msg="Failed to allocate and map port 80-80: Bind for fc00:1337:1337:1337:1337:1337:1337:1337:80 failed: port is already allocated"
Dec 19 12:04:11 mana dockerd[384483]: time="2020-12-19T12:04:11.682239315+11:00" level=error msg="ed24834d6f45bbaf301c8243f51a942bb82d81675c70481f68d64cf14f8f005e cleanup: failed to delete container from containerd: no such container"
Dec 19 12:04:11 mana dockerd[384483]: time="2020-12-19T12:04:11.693431409+11:00" level=error msg="Handler for POST /v1.41/containers/nginx-test/restart returned error: Cannot restart container nginx-test: driver failed programming external connectivity on endpoint nginx-test (9a6f7a503369a88aa1d200050e4a82ba09e755e48588f4000086c5abae6c3710): Bind for fc00:1337:1337:1337:1337:1337:1337:1337:80 failed: port is already allocated"
```

**Describe the results you expected:**

I expect that the container would have restarted successfully with the appropriate port being published on the specified IPv6 address.

**Additional information you deem important (e.g. issue happens only occasionally):**

I cannot reproduce the issue if publishing a port to a hosts IPv4 address, for example:
1. `docker run -d --name nginx-test -p 10.10.10.4:80:80 nginx`
2. `docker restart nginx-test`

The above commands will run successfully.

Downgrading to `19.03.14` provides a temporary workaround as this issue does not exist in the older version.
While my default runtime is set to `crun` I also went back to the Docker defaults with `runc` and I was able to reproduce the issue.

**Output of `docker version`:**

```
Client:
 Version:           20.10.1
 API version:       1.41
 Go version:        go1.15.6
 Git commit:        831ebeae96
 Built:             Tue Dec 15 22:25:01 2020
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server:
 Engine:
  Version:          20.10.1
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.15.6
  Git commit:       f0014860c1
  Built:            Tue Dec 15 22:24:28 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b.m
 crun:
  Version:          0.16
  GitCommit:        eb0145e5ad4d8207e84a327248af76663d4e50dd
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.5.1-tp-docker)

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 32
 Server Version: 20.10.1
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: crun io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: crun
 Init Binary: docker-init
 containerd version: 269548fa27e0089a8b8278fc4fc781d7f65a939b.m
 runc version: eb0145e5ad4d8207e84a327248af76663d4e50dd
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.9.14-arch1-1
 Operating System: Archcraft
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 31.32GiB
 Name: mana
 ID: WSAG:GWTA:H6MH:IIOR:M5TO:DMEQ:6HMW:BJ25:VB3W:ERGY:ORNA:3RGQ
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Registry Mirrors:
  http://nerv:5000/
 Live Restore Enabled: false

WARNING: No blkio weight support
WARNING: No blkio weight_device support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Not applicable.

### PR DESCRIPTION
**Pull request by arkodg**
Brings in https://github.com/moby/libnetwork/pull/2604
fixes https://github.com/moby/moby/issues/41822 "Upgrade to 20.10 fails to release ports published to IPv6 literal addresses"

### CODE REVIEW

**Review by AkihiroSuda — APPROVED** [review]

**Comment by nightah** [comment]
Am I correct in understanding that this should resolve #41822?

**Comment by arkodg** [comment]
yes @nightah this PR should solve your issue

**Comment by arkodg** [comment]
PTAL @tiborvass @thaJeztah

**Review by tiborvass — APPROVED** [review]

**Comment by gamma** [comment]
The solution of this issue has created a problem - at least on my -side: creating a container with option `-P` or `-p 80` assigns both, IPv4 __and__ IPv6 ports of the host to the container. Previously (version 20.10.1) it only assigned an IPv4 address.

On occasions it assigns these automatic ports to different port numbers on IPv4 and IPv6 - see https://forums.docker.com/t/why-different-port-number-is-mapped-on-ipv4-and-ipv6-address-for-same-expose-port/109795

This is a problem - at least when using the classic swarm - which seems to think that the IPv6 port is an IPv4 port. When the ports are not in sync it occurs, that, when several container with this issue are created, the IPV4 ports overlap on the swarm endpoint.

Long story short: before 20.10.2 there was no automatic IPv6 assignment using -P or -p 80. Now there is. Is that intended behavior?  If so: how can it be disabled without loosing IPv6 inside the container?

**Comment by thaJeztah** [comment]
Older releases implicitly bound ports on both IPv4 and IPv6. 20.10 includes that information in the API. 

> This is a problem - at least when using the classic swarm - which seems to think that the IPv6 port is an IPv4 port. 

That's unlikely to be fixed, with classic swarm being deprecated. 

for the IPv4 and IPv6 ports not being the same when using random / ephemeral port assignment, there's this ticket; https://github.com/moby/moby/issues/42442 (feel free to participate in the discussion there; contributions welcome as well of course 😅)


Not sure if it helps as a workaround, but if you know which ports should be mapped, you can use `0.0.0.0` as the host IP address, and have an ephemeral/random port assigned (leave out the `<host-port>` in the `<ip-address>:<host-port>:<container-port>` format). This should make the port only published on IPv4;

```bash
docker run -d -p 0.0.0.0::80 --name foobar nginx:alpine

docker port foobar 80
0.0.0.0:49155
```

**Comment by gamma** [comment]
Thanks for the elaborate explanation! I did not know about the `-p 0.0.0.0::80` syntax even though I'm using docker since its beginning 🤣 - it might help me indeed.

I'll also check the mentioned issue!

**Comment by thaJeztah** [comment]
Yeah, the `-p` _shorthand_ syntax is unfortunately quite complex. The `-P` / `--publish-all` option can be "convenient", but in general, I recommend being explicit about which container ports must be published. The `-P` option automatically creates a port-mapping based on metadata of the container (`EXPOSE`). The `EXPOSE` metadata is just _metadata_, and does not guarantee that something is actually listening on those ports (e.g., the `nginx` image has both `443` and `80` as `expose`, but by default only listens on port `80`.

Being explicit is more work, but makes sure that you _know_ what ports will be published, and keeps you in control whether or not a port should be published. Publishing a port makes it publicly accessible and in many situations (e.g. mysql port `3306`) that may not be desirable.
